native/effektio/src/platform/android.rs

```diff
@@ -12,19 +23,137 @@ pub async fn new_client_config(base_path: String, home: String) -> Result<Client
 
+    let log_level = match filter {
+        Some(ref filter) => FilterBuilder::new().parse(&filter).build(),
+        None => FilterBuilder::new().build(),
+    };
+
     let mut log_config = Config::default()
-        .with_min_level(Level::Trace)
-        .with_tag("effektio-sdk");
+        .with_max_level(LevelFilter::Trace)
+        .with_tag(app_name.as_str());
     if let Some(filter) = filter {
-        log_config = log_config.with_filter(FilterBuilder::new().parse(&filter).build())
+        log_config = log_config.with_filter(FilterBuilder::new().parse(&filter).build());
     }
+    let console_logger = LoggerWrapper::new(log_config).cloned_boxed_logger();
 
-    android_logger::init_once(log_config);
+    let mut path = PathBuf::from(log_dir.as_str());
+    path.push("app_");
+
+    let (level, dispatch) = fern::Dispatch::new()
+        .format(|out, message, record| {
+            out.finish(format_args!(
+                "{}[{}][{}] {}",
+                chrono::Local::now().format("[%Y-%m-%d][%H:%M:%S]"),
+                record.target(),
+                record.level(),
+                message
+            ))
+        })
+        // Only log messages Info and above
+        .level(log_level.filter())
+        // Output to console
+        .chain(console_logger)
+        // Output to file
+        .chain(fern::Manual::new(path, "%Y-%m-%d_%H-%M-%S%.f.log"))
+        .into_dispatch_with_arc();
+
+    if level == log::LevelFilter::Off {
+        log::set_boxed_logger(Box::new(native::NopLogger)).unwrap();
+    } else {
+        log::set_boxed_logger(Box::new(dispatch.clone())).unwrap();
+    }
+    log::set_max_level(level);
+
+    unsafe {
+        FILE_LOGGER = Some(dispatch);
+    }
 
     Ok(())
 }
+
+/// Wrapper for our console which acts as the actual logger.
+#[derive(Clone)]
+struct LoggerWrapper(Arc<Mutex<AndroidLogger>>);
+
+impl LoggerWrapper {
+    fn new(config: Config) -> Self {
+        let logger = AndroidLogger::new(config);
+        LoggerWrapper(Arc::new(Mutex::new(logger)))
+    }
+
+    fn cloned_boxed_logger(&self) -> Box<dyn Log> {
+        Box::new(self.clone())
+    }
+}
+
+impl Log for LoggerWrapper {
+    fn enabled(&self, metadata: &Metadata) -> bool {
+        metadata.level() <= Level::Info
+    }
+
+    fn log(&self, record: &Record) {
+        if self.enabled(record.metadata()) {
+            self.0.lock().unwrap().log(record);
+        }
+    }
+
+    fn flush(&self) {}
+}
+
+pub async fn report_bug(
+    url: String,
+    username: String,
+    password: Option<String>,
+    app_name: String,
+    version: String,
+    text: String,
+    label: String,
+    with_log: bool,
+) -> Result<bool> {
+    let mut form = Form::new()
+        .text("text", text)
+        .text("user_agent", "Mozilla/0.9")
+        .text("app", app_name)
+        .text("version", version)
+        .text("label", label);
+    if with_log {
+        unsafe {
+            if let Some(dispatch) = &FILE_LOGGER {
```

Comment:
What if this isn't set? We should probably `Err` here, rather than hiding the bug that our bug reporting is broken.

Reply:
If we use `Err`, app will crash.
I think it is preferable to use `Ok(false)` than `Err`.
```rust
    if with_log {
        match &*FILE_LOGGER.lock().unwrap() {
            Some(dispatch) => {
                let res = dispatch.rotate();
                for output in res.iter() {
                    match output {
                        Some((old_path, new_path)) => {
                            let log_path = old_path.canonicalize()?.to_string_lossy().to_string();
                            let file = fs::read(log_path)?;
                            let filename =
                                old_path.file_name().unwrap().to_string_lossy().to_string();
                            let file_part = Part::bytes(file)
                                .file_name(filename)
                                .mime_str("text/plain")?;
                            form = form.part("log", file_part);
                            break;
                        }
                        None => {}
                    }
                }
            }
            None => {
                return Ok(false);
            }
        }
    }
```

Reply:
Well, it crashed because the outer layer isn't handling the `Err` correctly, isn't it? It should probably `catch`, while we here still `Err` (which shouldn't be the case usually anyways).

Reply:
What about this code?
I replaced `Ok(false)` with `bail!()`.
```rust
    if with_log {
        match &*FILE_LOGGER.lock().unwrap() {
            Some(dispatch) => {
                let res = dispatch.rotate();
                for output in res.iter() {
                    match output {
                        Some((old_path, new_path)) => {
                            let log_path = old_path.canonicalize()?.to_string_lossy().to_string();
                            let file = fs::read(log_path)?;
                            let filename =
                                old_path.file_name().unwrap().to_string_lossy().to_string();
                            let file_part = Part::bytes(file)
                                .file_name(filename)
                                .mime_str("text/plain")?;
                            form = form.part("log", file_part);
                            break;
                        }
                        None => {}
                    }
                }
            }
            None => {
                return bail!("You didn't set up file logger.");
            }
        }
    }
```

Reply:
Fixed by https://github.com/acterglobal/a3/pull/488/commits/586bec5b92bf6d22a591b1a77bfcc25062b527b7